tests/LtiAssignmentsGradesServiceTest.php

```diff
@@ -76,6 +119,131 @@ public function testItGetsSingleLineItemWithReadonlyScope()
         $this->assertEquals($expected, $result);
     }
 
+    public function testItPutsAGrade()
+    {
+        $serviceData = [
+            'scope' => [LtiConstants::AGS_SCOPE_SCORE],
+            'lineitems' => 'https://canvas.localhost/api/lti/courses/8/line_items',
+        ];
+        $service = new LtiAssignmentsGradesService($this->connector, $this->registration, $serviceData);
+
+        $lineItem = new LtiLineitem([
+            'id' => 'testId',
+        ]);
+
+        $expected = [
+            'scoreGiven' => 10,
+            'scoreMaximum' => 15,
+        ];
+        $grade = new LtiGrade($expected);
+        $this->connector->shouldReceive('makeServiceRequest')
+            ->once()->andReturn($expected);
+
+        $result = $service->putGrade($grade, $lineItem);
+
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testItFindsALineItem()
+    {
+        $ltiLineitemData = [
+            'id' => 'testId',
+        ];
+
+        $serviceData = [
+            'scope' => [LtiConstants::AGS_SCOPE_LINEITEM],
+            'lineitems' => 'https://canvas.localhost/api/lti/courses/8/line_items',
+        ];
+
+        $lineItem = new LtiLineitem($ltiLineitemData);
+        $service = new LtiAssignmentsGradesService($this->connector, $this->registration, $serviceData);
+
+        $response = [$ltiLineitemData];
+
+        $this->connector->shouldReceive('getAll')
+            ->once()->andReturn($response);
+
+        $result = $service->findLineItem($lineItem);
+
+        $this->assertEquals($lineItem, $result);
+    }
+
+    public function testItFindsNoLineItem()
+    {
+        $ltiLineitemData = [
+            'id' => 'testId',
+        ];
+
+        $serviceData = [
+            'scope' => [LtiConstants::AGS_SCOPE_LINEITEM],
+            'lineitems' => 'https://canvas.localhost/api/lti/courses/8/line_items',
+        ];
+
+        $lineItem = new LtiLineitem($ltiLineitemData);
+        $service = new LtiAssignmentsGradesService($this->connector, $this->registration, $serviceData);
+
+        $this->connector->shouldReceive('getAll')
+            ->once()->andReturn([]);
+
+        $result = $service->findLineItem($lineItem);
```

Comment:
This is a totally unrelated question but why dont we compare lineitem `id` in `LtiAssignmentsGradeService::isMatchingLineitem`?

Reply:
Great question. For background, here's the discussion about it on the PR that introduced it: https://github.com/packbackbooks/lti-1-3-php-library/pull/35#discussion_r741394468

Ok, it took me a while to figure out, but I think I remember now finally. It's because when we're syncing grades, our app doesn't necessarily know if the line item exists in the LMS yet. As such the ID could be empty.

Both the `getGrades()` and `putGrade()` methods call `ensureLineItemExists()` which calls `findOrCreateLineItem()` which calls `findLineItem()` which calls `isMatchingLineItem()`.

In the case where we call `get|putGrade[s]()` and we need to create the line item still, we won't have an ID. (We will, however, have the tag, resource ID, and resource link ID.) Hence why we don't match on ID.

Tho I suppose we could add a `$newLineItem->getId() == ($lineitem['id'] ?? null) ||` to the beginning of `isMatchingLineitem()`